htlcswitch/hop/error_encryptor.go

```diff
@@ -79,22 +96,80 @@ type ErrorEncrypter interface {
 // result, all errors handled are themselves wrapped in layers of onion
 // encryption and must be treated as such accordingly.
 type SphinxErrorEncrypter struct {
-	*sphinx.OnionErrorEncrypter
+	encrypter interface{}
 
 	EphemeralKey *btcec.PublicKey
+	router       *sphinx.Router
+
+	createdAt time.Time
 }
 
-// NewSphinxErrorEncrypter initializes a blank sphinx error encrypter, that
-// should be used to deserialize an encoded SphinxErrorEncrypter. Since the
-// actual encrypter is not stored in plaintext while at rest, reconstructing the
-// error encrypter requires:
+// NewSphinxErrorEncrypterUninitialized initializes a blank sphinx error
+// encrypter, that should be used to deserialize an encoded
+// SphinxErrorEncrypter. Since the actual encrypter is not stored in plaintext
+// while at rest, reconstructing the error encrypter requires:
 //  1. Decode: to deserialize the ephemeral public key.
 //  2. Reextract: to "unlock" the actual error encrypter using an active
 //     OnionProcessor.
-func NewSphinxErrorEncrypter() *SphinxErrorEncrypter {
+func NewSphinxErrorEncrypterUninitialized() *SphinxErrorEncrypter {
 	return &SphinxErrorEncrypter{
-		OnionErrorEncrypter: nil,
-		EphemeralKey:        &btcec.PublicKey{},
+		EphemeralKey: &btcec.PublicKey{},
+		createdAt:    time.Now(),
+	}
+}
+
+func NewSphinxErrorEncrypter(router *sphinx.Router,
+	ephemeralKey *btcec.PublicKey) (*SphinxErrorEncrypter, error) {
+
+	encryptor, err := sphinx.NewOnionErrorEncrypter(
+		router, ephemeralKey,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SphinxErrorEncrypter{
+		encrypter:    encryptor,
+		router:       router,
+		EphemeralKey: ephemeralKey,
+		createdAt:    time.Now(),
+	}, nil
+}
+
+func (s *SphinxErrorEncrypter) UpgradeToFatErrors() error {
+	encryptor, err := sphinx.NewOnionFatErrorEncrypter(
+		s.router, s.EphemeralKey, FatErrorStructure,
```

Comment:
It may be better to not use a hard-coded structure here and instead let the sender specify the structure via the onion (https://github.com/lightning/bolts/pull/1044#discussion_r1069623771).

Reply:
Implemented this.